--- src/cpu/xbyak/xbyak_util.h	2019-02-12 20:59:33.810111340 -0800
+++ src/cpu/xbyak/xbyak_util_patched.h	2019-02-12 21:00:16.865359245 -0800
@@ -183,7 +183,7 @@
 					* (data[2] + 1);
 				if (cacheType == DATA_CACHE && smt_width == 0) smt_width = nb_logical_cores;
 				assert(smt_width != 0);
-				cores_sharing_data_cache[data_cache_levels] = nb_logical_cores / smt_width;
+				cores_sharing_data_cache[data_cache_levels] = (std::max)(nb_logical_cores / smt_width, 1u);
 				data_cache_levels++;
 			}
 		}
